module/os/freebsd/spl/spl_uio.c

```diff
@@ -105,3 +109,271 @@ zfs_uio_fault_move(void *p, size_t n, zfs_uio_rw_t dir, zfs_uio_t *uio)
 	ASSERT3U(zfs_uio_rw(uio), ==, dir);
 	return (vn_io_fault_uiomove(p, n, GET_UIO_STRUCT(uio)));
 }
+
+/*
+ * Check if the uio is page-aligned in memory.
+ */
+boolean_t
+zfs_uio_page_aligned(zfs_uio_t *uio)
+{
+	const struct iovec *iov = GET_UIO_STRUCT(uio)->uio_iov;
+
+	for (int i = zfs_uio_iovcnt(uio); i > 0; iov++, i--) {
+		unsigned long addr = (unsigned long)iov->iov_base;
+		size_t size = iov->iov_len;
+		if ((addr & (PAGE_SIZE - 1)) || (size & (PAGE_SIZE - 1))) {
+				return (B_FALSE);
+		}
+	}
+
+	return (B_TRUE);
+}
+
+#if __FreeBSD_version < 1300050
+static void
+zfs_uio_set_pages_to_stable(zfs_uio_t *uio)
+{
+	vm_object_t obj;
+
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	ASSERT3U(uio->uio_dio.npages, >, 0);
+
+	obj = uio->uio_dio.pages[0]->object;
+	zfs_vmobject_wlock(obj);
+	for (int i = 0; i < uio->uio_dio.npages; i++) {
+		vm_page_t page = uio->uio_dio.pages[i];
+
+		ASSERT3P(page, !=, NULL);
+		vm_page_sbusy(page);
+		MPASS(page == PHYS_TO_VM_PAGE(VM_PAGE_TO_PHYS(page)));
+		if (page->object != obj) {
+			zfs_vmobject_wunlock(obj);
+			obj = page->object;
+			zfs_vmobject_wlock(obj);
+		}
+		pmap_remove_write(page);
+	}
+	zfs_vmobject_wunlock(obj);
+}
+
+static void
+zfs_uio_release_stable_pages(zfs_uio_t *uio)
+{
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	for (int i = 0; i < uio->uio_dio.npages; i++) {
+		vm_page_t page = uio->uio_dio.pages[i];
+
+		ASSERT3P(page, !=, NULL);
+		ASSERT(vm_page_sbusied(page));
+		vm_page_sunbusy(page);
+	}
+}
+
+#else
+
+static void
+zfs_uio_set_pages_to_stable(zfs_uio_t *uio)
+{
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	ASSERT3U(uio->uio_dio.npages, >, 0);
+
+	for (int i = 0; i < uio->uio_dio.npages; i++) {
+		vm_page_t page = uio->uio_dio.pages[i];
+		ASSERT3P(page, !=, NULL);
+
+		MPASS(page == PHYS_TO_VM_PAGE(VM_PAGE_TO_PHYS(page)));
+		vm_page_busy_acquire(page, VM_ALLOC_SBUSY);
+		pmap_remove_write(page);
```

Comment:
So here we're removing managed writeable mappings of each page since we want their contents to remain stable. But this isn't really sufficient, since the kernel can modify those pages via the direct map, which isn't affected by pmap_remove_write().

For example, suppose a process sneakily tries to write and read a buffer at the same time, using a file that is opened for direct I/O. Suppose ZFS handles reading first; it'll use vm_fault_quick_hold_pages() to look up and wire the pages backing the buffer; then, it'll do the same thing again for the write, after which it'll call pmap_remove_write() on each page. Now the first read() call still has the ability to modify the pages, because ZFS/uiomove() is using the direct map to access their contents, rather than the user mapping that is removed by pmap_remove_write().

To really secure exclusive access to the page, we need to ensure that no other wiring references exist to any page in the range. vm_page_try_remove_write() will do this. A stumbling block there is that the page is already wired by the current thread, so vm_page_try_remove_write() will always fail.

It's possible to handle this problem, but I'm not sure what invariants we're required to preserve here. Is it at all acceptable for the backing pages to be stable after they're made "stable", or would that result in data corruption? Can we have a fallback path where the contents of the pages are copied into a private buffer if it's not possible to exclude writers? I think the latter will be necessary on FreeBSD.

Reply:
@markjdb could this situation of a a process doing a read and write at the same time even be possible? In `zfs_vnops.c` range locks are acquired, so I am not certain that a process could read from the same range that is being written to. However, this may just be a misunderstanding on my part. Are you instead saying that shared pages between the processes are read/writing from different regions of the file? If so, would that require a `mmap()` operation? We do bail on doing any direct I/O if the region of file is memory mapped. Just trying to understand this more. I am no way an expert on how FreeBSD would handle this and I appreciate the help with resolving this issue.

Reply:
ZFS range locks would just prevent simultaneous reading and writing from the same range of a file, right? If so, that's not enough to prevent modifications. To state it another way, the problem I'm describing is that pmap_remove_write() really only prevents userspace from modifying the pages backing the I/O buffer; it doesn't prevent the kernel from performing I/O to/from the buffer. That I/O could be from a filesystem, a pipe, a socket, etc..

For kernel I/O, the thread performing the operation need only hold a wiring of the page(s). The existing of a wiring of the target page means that something else might be modifying the contents of the page. The vm_page_try_remove_write() function atomically removes all writeable mappings while ensuring that no new wirings can be created. It fails if a wiring already exists. I think that we'll want to use that here and bail if the operation fails (i.e., fall back to copying), but I'm not sure exactly what goes wrong if a thread somehow manages to modify a page while ZFS assumes that it's stable.

Reply:
I'm using dtrace to figure out where things are failing, but understanding why is a bit more challenging.

> `pmap_extract_and_hold()` is failing and then `vm_fault()` fails with ENOENT

Slight correction: it's not `ENOENT`, it's `KERN_PROTECTION_FAILURE` which seems to come from here:

```
int
vm_fault(vm_map_t map, vm_offset_t vaddr, vm_prot_t fault_type,
    int fault_flags, vm_page_t *m_hold)
{
        struct faultstate fs;
        int ahead, behind, faultcount, rv;
        enum fault_status res;
        bool hardfault;

        VM_CNT_INC(v_vm_faults);

        if ((curthread->td_pflags & TDP_NOFAULTING) != 0)
                return (KERN_PROTECTION_FAILURE);
```

That makes this snippet from `vm_fault_quick_hold_pages()` seem a bit silly, since regardless of `prot` we're going to get an error from `vm_fault()` because of `TDP_NOFAULTING`. I guess we could fail faster by setting `VM_PROT_QUICK_NOFAULT` though:

```
                /*
                 * One or more pages could not be held by the pmap.  Either no
                 * page was mapped at the specified virtual address or that
                 * mapping had insufficient permissions.  Attempt to fault in
                 * and hold these pages.
                 *
                 * If vm_fault_disable_pagefaults() was called,
                 * i.e., TDP_NOFAULTING is set, we must not sleep nor
                 * acquire MD VM locks, which means we must not call
                 * vm_fault().  Some (out of tree) callers mark
                 * too wide a code area with vm_fault_disable_pagefaults()
                 * already, use the VM_PROT_QUICK_NOFAULT flag to request
                 * the proper behaviour explicitly.
                 */
                if ((prot & VM_PROT_QUICK_NOFAULT) != 0 &&
                    (curthread->td_pflags & TDP_NOFAULTING) != 0)
                        goto error;
                for (mp = ma, va = addr; va < end; mp++, va += PAGE_SIZE)
                        if (*mp == NULL && vm_fault(map, va, prot,
                            VM_FAULT_NORMAL, mp) != KERN_SUCCESS)
                                goto error;
```

On FreeBSD 14 we never even call `vm_fault()` because `pmap_extract_and_hold()` succeeds. I'm still not sure  why that is, I haven't looked much at pmap yet.

Using `vm_page_try_remove_write()` and handling errors appropriately sounds like a good idea, but we're failing in `zfs_uio_get_dio_pages_impl()` which is called before `zfs_uio_set_pages_to_stable()`, so it's not obvious how that would make a difference.

Reply:
I've tested using `vm_page_try_remove_write()` and returning EFAULT if it fails, but as expected it does not solve the problem of `vm_fault_quick_hold_pages()` failing.